### PR TITLE
Parallelize sort using libstdc++ parallel mode

### LINKED ISSUE #149977: UNSTABLE inductor / linux-jammy-cpu-py3.9-gcc11-inductor / test (cpu_inductor_torchbench)
> For example, DISABLED pull / win-vs2022-cpu-py3 / test (default). Once
> created, the job will be disabled within 15 minutes. You can check the
> list of disabled jobs at https://ossci-metrics.s3.amazonaws.com/disabled-jobs.json

> If you need to get this out ASAP instead of waiting for 15 minutes,
> you can manually trigger the workflow at https://github.com/pytorch/test-infra/actions/workflows/update_disabled_tests.yml
> once the issue is created to update the above JSON list right away.

> Noted: you need to have write access to PyTorch repo to disable CI
> jobs. The issue will be rejected otherwise.

## Reason
the test keeps failing in HUD

cc @seemethere @malfet @pytorch/pytorch-dev-infra @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by annop-w**
Fixes #149977, #149979, #150094.

Previously, #149505 used libstdc++ parallel mode by enabling `-D_GLIBCXX_PARALLEL`. However, mixing source files compiled with and without parallel mode can lead to undefined behavior (See https://gcc.gnu.org/onlinedocs/libstdc++/manual/parallel_mode_using.html) We switch to using the specific paralell sort from `<parallel/algorithm>` when compiled with GCC compiler. Note that use of `std::execution` policy has dependency on libtbb and we thus decide to avoid that.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @malfet @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01

### CODE REVIEW

**Comment by cyyever** [comment]
Can we use c++17 parallel sort?

**Comment by annop-w** [comment]
@cyyever That has dependency on tbb which we would like to avoid adding.

**Comment by cyyever** [comment]
@annop-w I see, but is it possible to use them for Clang and MSVC?

**Comment by annop-w** [comment]
@cyyever It works for clang and gcc. I have not tried MSVC. But again, there is dependency on tbb and if we would like to avoid adding libtbb dependency in PyTorch, then we aren't able to use the parallel execution policy.

File: aten/src/ATen/native/cpu/SortingKernel.cpp [hunk 1]
```diff
@@ -19,6 +19,9 @@
 #ifdef USE_FBGEMM
 #include <fbgemm/Utils.h>
 #endif
+#if defined(__GNUC__) && !defined(__clang__) && defined(_OPENMP)
```
Comment (cyyever) [inline aten/src/ATen/native/cpu/SortingKernel.cpp:22]: Is it better to use ``#if __has_include(<parallel/algorithm>) && defined(_OPENMP)`` ?

Comment (annop-w) [reply aten/src/ATen/native/cpu/SortingKernel.cpp]: That's a really good suggestion. Done 👍

**Comment by cyyever** [comment]
Can you provide benchmark numbers before and after changing?

File: aten/src/ATen/native/cpu/SortingKernel.cpp [hunk 2]
```diff
@@ -146,6 +149,25 @@ static inline void sort_kernel_impl(const value_accessor_t& value_accessor,
   auto composite_accessor = CompositeRandomAccessorCPU<
     value_accessor_t, indices_accessor_t
   >(value_accessor, indices_accessor);
+#if __has_include(<parallel/algorithm>) && defined(_OPENMP)
```
[the remark this answers is not in the document]
Comment (annop-w) [reply aten/src/ATen/native/cpu/SortingKernel.cpp:152]: I think that is a bit out of scope for this PR. It can be done easily in a separate PR though :)

**Comment by annop-w** [comment]
> Can you provide benchmark numbers before and after changing?

I see ~9.7x speedup with 16 threads on NEOVERSE V1 when sorting ~50000 elements.

**Comment by annop-w** [comment]
@malfet Could I please get a review ?

**Comment by annop-w** [comment]
@digantdesai Could you please help taking a look ? The checks passed and the failing one looks unrelated to me.

**Comment by annop-w** [comment]
@malfet Could you please review this one ? Thank you.

File: aten/src/ATen/native/cpu/SortingKernel.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (digantdesai) [reply aten/src/ATen/native/cpu/SortingKernel.cpp:152]: I think this is valid. let's figure out the sorting function first and just call it once?

[the remark this answers is not in the document]
Comment (annop-w) [reply aten/src/ATen/native/cpu/SortingKernel.cpp:152]: Done.

**Comment by annop-w** [comment]
@cyyever @digantdesai Could you please review again ? Thank you.

**Comment by cyyever** [comment]
@malfet Have a look because performance gain fro proper C++17 parallel is desirable.

**Comment by malfet** [comment]
> @malfet Have a look because performance gain fro proper C++17 parallel is desirable.

But this PR is not using c++17 parallel sort, but rather some g++ extension?

Also, all 3 issues that PR claims to fix has been closed, but I also struggle to understand how it could have fixed them in the first place

**Comment by malfet** [comment]
> @cyyever It works for clang and gcc. I have not tried MSVC. But again, there is dependency on tbb and if we would like to avoid adding libtbb dependency in PyTorch, then we aren't able to use the parallel execution policy.

@annop-w Can you please reference the doc? To the best of my knowledge functions which are part of the standard should rely only on  c++ runtime shipped with compiler.

If this is not the case, then I think one should write an in-house implementation using at::parallel primitive

**Comment by annop-w** [comment]
@malfet
Here is the doc for GNU https://gcc.gnu.org/onlinedocs/libstdc++/manual/parallel_mode_using.html. 
I will look for docs for Clang.

The 3 issues are closed because #149505 got reverted. I explained how this PR would also solve the issues here

> Previously, https://github.com/pytorch/pytorch/pull/149505 used libstdc++ parallel mode by enabling -D_GLIBCXX_PARALLEL. However, mixing source files compiled with and without parallel mode can lead to undefined behavior

**Comment by cyyever** [comment]
@annop-w @malfet For your reference, [libstdc++ document](https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2017) says 
```
 Note 3: The Parallel Algorithms have an external dependency on Intel TBB 2018 or later. 
If the <execution> header is included then -ltbb must be used to link to TBB. 
```
Currently, [libc++](https://libcxx.llvm.org/Status/Cxx17.html) has no full Parallelism TS support.

**Review by malfet — CHANGES_REQUESTED** [review]
Requesting changes because:
 - We should avoid using random compiler extensions in the codebase (especially when they work for ascending but not descending)
 - Previous attempt to enable libstdc++ enable parallel sort resulted in correctness issue and this PR adds no unit test nor provide a satisfactory explanation(fact that something might lead to undefined behavior, in my mind is not sufficient, my strong suspicious on what happened are results were unstable compared to sequential sort when there were duplicated indices) why this use of extension will not suffer from the same fate 
 - Last but not least, this is a performance optimization, but no script are provided that one can use to validate that

To conclude, if looks like if one wants to enable parallel sort, they could not rely on C++17 primitives, as their implementation at least on Linux is currently flawed, but instead should write something that relies on `at::parallel_for` primitive

**Comment by annop-w** [comment]
@malfet I am sorry but I do not agree with your arguments 

> We should avoid using random compiler extensions in the codebase (especially when they work for ascending but not descending)

This is not just some random extensions. It is just GNU version of C++ standard library which is also compatible with Clang. For Clang, the C++ std. library implementation can be selected with `-stdlib=` flag. The__gnu_parallel version has the same function signature and accepts a comparator, i.e. it works for both ascending and descending sort.

> Previous attempt to enable libstdc++ enable parallel sort resulted in correctness issue and this PR adds no unit test nor provide a satisfactory explanation(fact that something might lead to undefined behavior, in my mind is not sufficient, my strong suspicious on what happened are results were unstable compared to sequential sort when there were duplicated indices) why this use of extension will not suffer from the same fate

This is documented in the documentation as

> Note that the _GLIBCXX_PARALLEL define may change the sizes and behavior of standard class templates such as std::search, and therefore one can only link code compiled with parallel mode and code compiled without parallel mode if no instantiation of a container is passed between the two translation units. Parallel mode functionality has distinct linkage, and cannot be confused with normal mode symbols.

I validated the change in this PR with script in #150094 and it passed the accuracy test. So, tests already exist.

> Last but not least, this is a performance optimization, but no script are provided that one can use to validate that

Benchmarking script is already provided in #142391 and I paste it here again for reference
```
import torch
import torch.autograd.profiler as profiler

torch.manual_seed(0)

N = 50000
x = torch.randn(N, dtype=torch.float)

with profiler.profile(with_stack=True, profile_memory=False, record_shapes=True) as prof:
    for i in range(1000):
        _, _ = torch.sort(x)

print(prof.key_averages(group_by_input_shape=True).table(sort_by='self_cpu_time_total', row_limit=10))
```

> if one wants to enable parallel sort, they could not rely on C++17 primitives, as their implementation at least on Linux is currently flawed, but instead should write something that relies on at::parallel_for primitive

I do not think the implementation is flawed. If there exists a working parallel sort in the C++ std. library, why reinvent the wheel ?

File: aten/src/ATen/native/cpu/SortingKernel.cpp [hunk 3]
```diff
@@ -19,6 +19,9 @@
 #ifdef USE_FBGEMM
 #include <fbgemm/Utils.h>
 #endif
+#if __has_include(<parallel/algorithm>) && defined(_OPENMP)
+#include <parallel/algorithm>
```
Comment (malfet) [inline aten/src/ATen/native/cpu/SortingKernel.cpp:23]: Is there a C++ standard that says this header will be available in C++20? Or C++23? If not, please don't rely on this header

**Review by malfet — CHANGES_REQUESTED** [review]
> This is not just some random extensions. It is just GNU version of C++ standard library which is also compatible with Clang. For Clang, the C++ std. library implementation can be selected with `-stdlib=` flag. The__gnu_parallel version has the same function signature and accepts a comparator, i.e. it works for both ascending and descending sort.

It's not part of C++ standard, therefore we should not rely on that.

> > Previous attempt to enable libstdc++ enable parallel sort resulted in correctness issue and this PR adds no unit test nor provide a satisfactory explanation(fact that something might lead to undefined behavior, in my mind is not sufficient, my strong suspicious on what happened are results were unstable compared to sequential sort when there were duplicated indices) why this use of extension will not suffer from the same fate
> 
> I validated the change in this PR with script in #150094 and it passed the accuracy test. So, tests already exist.

Can you please explain what changed in the approach between this PR and https://github.com/pytorch/pytorch/pull/149505 that make test pass now but fail previously?

> Benchmarking script is already provided in #142391 and I paste it here again for reference

Please add it to the PR description and share the numbers that you've observed before and after the change.

> I do not think the implementation is flawed. If there exists a working parallel sort in the C++ std. library, why reinvent the wheel ?

Didn't you mention that it works for ascending but not descending case (though I see you have a workaround for it now)

**Comment by annop-w** [comment]
We will be looking at implementing a solution based on `aten::parallel_for`.
